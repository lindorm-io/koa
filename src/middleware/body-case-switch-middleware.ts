import { IKoaAppContext, TNext } from "../typing";
import { camelKeys, snakeKeys } from "@lindorm-io/core";

export const bodyCaseSwitchMiddleware = async (ctx: IKoaAppContext, next: TNext): Promise<void> => {
  ctx.request.body = camelKeys(ctx.request.body) || {};

  await next();

  ctx.body = snakeKeys(ctx.body);
};
