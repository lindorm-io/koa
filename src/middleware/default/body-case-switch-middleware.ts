import type { KoaContext, Middleware } from "../../typing";
import { camelKeys, snakeKeys } from "@lindorm-io/core";

export const bodyCaseSwitchMiddleware: Middleware<KoaContext> = async (ctx, next): Promise<void> => {
  ctx.request.body = ctx.request.body ? camelKeys(ctx.request.body) : {};

  await next();

  ctx.body = ctx.body ? snakeKeys(ctx.body) : ctx.body;
};
