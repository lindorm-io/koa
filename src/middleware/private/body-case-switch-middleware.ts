import { KoaContext, Middleware } from "../../typing";
import { camelKeys, isObjectStrict, snakeKeys } from "@lindorm-io/core";

export const bodyCaseSwitchMiddleware: Middleware<KoaContext> = async (ctx, next): Promise<void> => {
  ctx.request.body = isObjectStrict(ctx.request.body) ? camelKeys(ctx.request.body) : {};

  await next();

  ctx.body = isObjectStrict(ctx.body) ? snakeKeys(ctx.body) : ctx.body;
};
