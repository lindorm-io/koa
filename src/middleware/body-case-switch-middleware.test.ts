import { bodyCaseSwitchMiddleware } from "./body-case-switch-middleware";

describe("body-case-switch-middleware.ts", () => {
  const array = ["array"];
  const date = new Date();
  const error = new Error("error");
  const string = "string";

  let ctx: any;
  let next: any;

  beforeEach(() => {
    ctx = {
      request: {
        body: {
          camelCase: "camelCase",
          PascalCase: "PascalCase",
          snake_case: "snake_case",
          array,
          date,
          error,
          string,
        },
      },
      body: {
        camelCase: "camelCase",
        PascalCase: "PascalCase",
        snake_case: "snake_case",
        array,
        date,
        error,
        string,
      },
    };

    next = () => Promise.resolve();
  });

  test("should transform all incoming body to camelCase", async () => {
    await expect(bodyCaseSwitchMiddleware(ctx, next)).resolves.toBe(undefined);
    expect(ctx.request.body).toStrictEqual({
      camelCase: "camelCase",
      pascalCase: "PascalCase",
      snakeCase: "snake_case",
      array,
      date,
      error,
      string,
    });
  });

  test("should transform all outgoing body to snake_case", async () => {
    await expect(bodyCaseSwitchMiddleware(ctx, next)).resolves.toBe(undefined);
    expect(ctx.body).toStrictEqual({
      camel_case: "camelCase",
      pascal_case: "PascalCase",
      snake_case: "snake_case",
      array,
      date,
      error,
      string,
    });
  });

  test("should not fail if there is no body", async () => {
    ctx = { request: { body: undefined }, body: undefined };

    await expect(bodyCaseSwitchMiddleware(ctx, next)).resolves.toBe(undefined);
    expect(ctx.request.body).toStrictEqual({});
    expect(ctx.body).toStrictEqual({});
  });
});
