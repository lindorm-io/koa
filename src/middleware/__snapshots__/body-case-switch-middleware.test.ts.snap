// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`bodyCaseSwitchMiddleware should not fail if there is no body 1`] = `Object {}`;

exports[`bodyCaseSwitchMiddleware should not fail if there is no body 2`] = `undefined`;

exports[`bodyCaseSwitchMiddleware should transform all incoming body to camelCase 1`] = `
Object {
  "array": Array [
    "array",
  ],
  "camelCase": "camelCase",
  "date": 2020-01-01T07:00:00.000Z,
  "error": [Error: error],
  "pascalCase": "PascalCase",
  "snakeCase": "snake_case",
  "string": "string",
}
`;

exports[`bodyCaseSwitchMiddleware should transform all outgoing body to snake_case 1`] = `
Object {
  "array": Array [
    "array",
  ],
  "camel_case": "camelCase",
  "date": 2020-01-01T07:00:00.000Z,
  "error": [Error: error],
  "pascal_case": "PascalCase",
  "snake_case": "snake_case",
  "string": "string",
}
`;
